beacon_node/http_api/src/block_id.rs

```diff
@@ -310,12 +324,48 @@ impl BlockId {
             }
             None => blob_sidecar_list,
         };
-        Ok((
-            block,
-            blob_sidecar_list_filtered,
-            execution_optimistic,
-            finalized,
-        ))
+
+        Ok(blob_sidecar_list_filtered)
+    }
+
+    fn get_blobs_from_data_columns<T: BeaconChainTypes>(
+        chain: &BeaconChain<T>,
+        root: Hash256,
+        blob_indices: Option<Vec<u64>>,
+        block: &SignedBlindedBeaconBlock<<T as BeaconChainTypes>::EthSpec>,
+    ) -> Result<BlobSidecarList<T::EthSpec>, Rejection> {
+        let column_indices = chain.store.get_data_column_keys(root).map_err(|e| {
+            warp_utils::reject::custom_server_error(format!(
+                "Error fetching data columns keys: {e:?}"
+            ))
+        })?;
+
+        let num_found_column_keys = column_indices.len();
+        let num_required_columns = chain.spec.number_of_columns / 2;
+        let is_blob_available = num_found_column_keys >= num_required_columns;
+
+        if is_blob_available {
+            let data_columns = column_indices
+                .into_iter()
+                .filter_map(
+                    |column_index| match chain.get_data_column(&root, &column_index) {
+                        Ok(Some(data_column)) => Some(Ok(data_column)),
+                        Ok(None) => None,
+                        Err(e) => Some(Err(warp_utils::reject::beacon_chain_error(e))),
+                    },
+                )
+                .collect::<Result<Vec<_>, _>>()?;
+
+            reconstruct_blobs(&chain.kzg, &data_columns, blob_indices, block).map_err(|e| {
+                warp_utils::reject::custom_server_error(format!(
+                    "Error reconstructing data columns: {e:?}"
+                ))
+            })
+        } else {
+            Err(warp_utils::reject::custom_server_error(
```

Comment:
Note: the status code for this hasn't been decided and spec may need to be updated.